sign/pending_requests.go

```diff
@@ -0,0 +1,185 @@
+func (rs *PendingRequests) Approve(id string, password string, verify verifyFunc) (hash gethcommon.Hash, err error) {
+	rs.log.Info("complete transaction", "id", id)
+	request, err := rs.tryLock(id)
+	if err != nil {
+		rs.log.Warn("can't process transaction", "err", err)
+		return hash, err
+	}
+
+	selectedAccount, err := verify(password)
+	if err != nil {
+		rs.complete(request, hash, err)
+		return hash, err
+	}
+
+	hash, err = request.completeFunc(selectedAccount)
+	rs.log.Info("finally completed transaction", "id", request.ID, "hash", hash, "err", err)
+
+	rs.complete(request, hash, err)
+
+	return hash, err
+}
+
+// Discard remove a signing request from the list of pending requests.
+func (rs *PendingRequests) Discard(id string) error {
+	request, err := rs.Get(id)
+	if err != nil {
+		return err
+	}
+
+	rs.complete(request, gethcommon.Hash{}, ErrSignReqDiscarded)
+	return nil
+}
+
+// Wait blocks until a request with a specified ID is completed (approved or discarded)
+func (rs *PendingRequests) Wait(id string, timeout time.Duration) Result {
+	request, err := rs.Get(id)
+	if err != nil {
+		return Result{Error: err}
+	}
+	for {
+		select {
+		case rst := <-request.result:
+			return rst
+		case <-time.After(timeout):
+			rs.complete(request, gethcommon.Hash{}, ErrSignReqTimedOut)
+		}
+	}
+}
+
+// Count returns number of currently pending requests
+func (rs *PendingRequests) Count() int {
+	rs.mu.RLock()
+	defer rs.mu.RUnlock()
+	return len(rs.requests)
+}
+
+// Has checks whether a pending request with a given identifier exists in the list
+func (rs *PendingRequests) Has(id string) bool {
+	rs.mu.RLock()
+	defer rs.mu.RUnlock()
+	_, ok := rs.requests[id]
+	return ok
+}
+
+// tryLock is used to avoid double-completion of the same request.
+// it returns a request instance if it isn't processing yet, returns an error otherwise.
+func (ts *PendingRequests) tryLock(id string) (*Request, error) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	if tx, ok := ts.requests[id]; ok {
+		if tx.locked {
+			return nil, ErrSignReqInProgress
+		}
+		tx.locked = true
+		return tx, nil
+	}
+	return nil, ErrSignReqNotFound
+}
+
+// complete removes the request from the list if there is no error or an error is non-transient
+func (rs *PendingRequests) complete(request *Request, hash gethcommon.Hash, err error) error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+
+	request.locked = false
+
+	go NotifyOnReturn(request.ID, request.Meta, request.context, err)
+
+	if err != nil && IsTransient(err) {
+		return nil
+	}
+
+	delete(rs.requests, request.ID)
+
+	// hash is updated only if err is nil, but transaction is not removed from a queue
+	result := Result{Error: err}
+	if err == nil {
+		result.Hash = hash
+	}
+
+	// sending result to the request in a separate goroutine to avoid deadlocks if
+	// both `Wait` and `Complete` are called from the same goroutine
+	// (they are in the tests, for instance).
+	go func() { request.result <- result }()
```

Comment:
Well, that is complicated, because the previous code did essentially everything through channelling. But I might try to remove that and see if your tests won't deadlock :)